starlite/contrib/repository/abc.py

```diff
@@ -0,0 +1,244 @@
+"""Data persistence interface."""
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from typing import TYPE_CHECKING, Any, Generic, TypeVar
+
+from .exceptions import NotFoundError
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from .types import FilterTypes
+
+
+__all__ = ("AbstractRepository",)
+
+T = TypeVar("T")
+RepoT = TypeVar("RepoT", bound="AbstractRepository")
+CollectionT = TypeVar("CollectionT")
+
+
+class AbstractRepository(Generic[T], metaclass=ABCMeta):
+    """Interface for persistent data interaction."""
+
+    model_type: type[T]
+    """Type of object represented by the repository."""
+    id_attribute = "id"
+    """Name of the primary identifying attribute on `model_type`."""
+
+    def __init__(self, **kwargs: Any) -> None:
+        """Repository constructors accept arbitrary kwargs."""
+        super().__init__(**kwargs)
+
+    @abstractmethod
+    async def add(self, data: T) -> T:
+        """Add `data` to the collection.
+
+        Args:
+            data: Instance to be added to the collection.
+
+        Returns:
+            The added instance.
+        """
+
+    @abstractmethod
+    async def add_many(self, data: Sequence[T]) -> Sequence[T]:
+        """Add multiple `data` to the collection.
+
+        Args:
+            data: Instances to be added to the collection.
+
+        Returns:
+            The added instances.
+        """
+
+    @abstractmethod
+    async def count(self, *filters: FilterTypes, **kwargs: Any) -> int:
+        """Get the count of records returned by a query.
+
+        Args:
+            *filters: Types for specific filtering operations.
+            **kwargs: Instance attribute value filters.
+
+        Returns:
+            The count of instances
+        """
+
+    @abstractmethod
+    async def delete(self, item_id: Any) -> T:
+        """Delete instance identified by `item_id`.
+
+        Args:
+            item_id: Identifier of instance to be deleted.
+
+        Returns:
+            The deleted instance.
+
+        Raises:
+            RepositoryNotFoundException: If no instance found identified by `item_id`.
+        """
+
+    @abstractmethod
+    async def get(self, item_id: Any, **kwargs: Any) -> T:
+        """Get instance identified by `item_id`.
+
+        Args:
+            item_id: Identifier of the instance to be retrieved.
+
+        Returns:
+            The retrieved instance.
+
+        Raises:
+            RepositoryNotFoundException: If no instance found identified by `item_id`.
+        """
+
+    @abstractmethod
+    async def get_one(self, **kwargs: Any) -> T:
+        """Get an instance if it exists or None.
```

Comment:
This doc string is not correct.

Reply:
or the return type is not correct.. actually, this method's place is unclear to me.. is this meant to be more like `first()` or smth?

Reply:
`get` requires an `id` field lookup.  `get_one` allows you to pass in an arbitrary set of  kwargs to use for the where clause.  It return one or an exception.

`get_one` and `get` could be combined.

Reply:
So, to answer this.  The docstring is incorrect here.  It should return the instance found with the kwarg filter or raise an exception.

Reply:
ok got ya